cmd/device-util/main.go

```diff
@@ -3,105 +3,41 @@
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
 )
 
-const usage = `device-util is a command line tool for querying devices present on Windows`
+const desc = `A stand-alone tool that replicates a limited subset of sc.exe and other tools.
+It can be used within WCOW uVMs and containers without requiring additional DLLs.`
 
 func main() {
-	app := cli.NewApp()
-	app.Name = "device-util"
-	app.Commands = []cli.Command{
-		queryChildrenCommand,
-		readObjDirCommand,
+	app := &cli.App{
+		Name:        "device-util",
+		Usage:       "tool for managing devices, drivers, and services on Windows",
+		Description: desc,
+		Commands: []*cli.Command{
+			driverCommand,
+			serviceCommand,
+			queryChildrenCommand,
+			readObjDirCommand,
+		},
 	}
-	app.Usage = usage
 
 	if err := app.Run(os.Args); err != nil {
-		fmt.Fprintln(os.Stderr, err)
-		os.Exit(1)
+		log.New(os.Stderr, "", 0).Fatal(err)
 	}
 }
 
-const (
-	parentIDFlag = "parentID"
-	propertyFlag = "property"
-	objDirFlag   = "dir"
-
-	locationProperty = "location"
-	idProperty       = "id"
-
-	globalNTPath = "\\Global??"
-)
-
-var readObjDirCommand = cli.Command{
-	Name:  "obj-dir",
-	Usage: "outputs contents of a NT object directory",
-	Flags: []cli.Flag{
-		cli.StringFlag{
-			Name:  objDirFlag,
-			Usage: "Optional: Object directory to query. Defaults to the global object directory.",
-		},
-	},
-	Action: func(context *cli.Context) error {
-		dir := globalNTPath
-		if context.IsSet(objDirFlag) {
-			dir = context.String(objDirFlag)
-		}
-		entries, err := winobjdir.EnumerateNTObjectDirectory(dir)
-		if err != nil {
-			return err
-		}
-		formatted := strings.Join(entries, ",")
-		fmt.Fprintln(os.Stdout, formatted)
-		return nil
-	},
-}
-
-var queryChildrenCommand = cli.Command{
-	Name:  "children",
-	Usage: "queries for given devices' children on the system",
-	Flags: []cli.Flag{
-		cli.StringFlag{
-			Name:  parentIDFlag,
-			Usage: "Required: Parent device's instance IDs. Comma separated string.",
-		},
-		cli.StringFlag{
-			Name:  propertyFlag,
-			Usage: "Either 'location' or 'id', default 'id'. String indicating a property to query devices for.",
-		},
-	},
-	Action: func(context *cli.Context) error {
-		if !context.IsSet(parentIDFlag) {
-			return errors.New("`children` command must specify at least one parent instance ID")
-		}
-		csParents := context.String(parentIDFlag)
-		parents := strings.Split(csParents, ",")
-
-		children, err := windevice.GetChildrenFromInstanceIDs(parents)
-		if err != nil {
-			return err
-		}
-
-		property := idProperty
-		if context.IsSet(propertyFlag) {
-			property = context.String(propertyFlag)
-		}
-		if property == locationProperty {
-			children, err = windevice.GetDeviceLocationPathsFromIDs(children)
-			if err != nil {
-				return err
-			}
-		}
-		formattedChildren := strings.Join(children, ",")
-		fmt.Fprintln(os.Stdout, formattedChildren)
-		return nil
-	},
+func verifyElevated(ctx *cli.Context) error {
+	if !winapi.IsElevated() {
```

Comment:
I assume we tested that this works properly in the UVM?

Reply:
yup, execs in the uVM run as admin

Reply:
I was asking because I figure we actually run as SYSTEM or something similar in the UVM, rather than as an administrator or elevated account. But if it works with testing that sounds good to me.

Reply:
oh, yeah, it is system i think
i am not too sure on the details, but windows says the process token has an elevated token, so it shouldn't break